aws_lambda_powertools/utilities/data_masking/base.py

```diff
@@ -26,7 +26,72 @@
 
 logger = logging.getLogger(__name__)
 
-
+def prepare_data(data: Any, _visited: set[int] | None = None) -> Any:
+    """
+    Recursively convert complex objects into dictionaries (or simple types) so that they can be
+    processed by the data masking utility. This function handles:
+
+    - Dataclasses (using dataclasses.asdict)
+    - Pydantic models (using model_dump)
+    - Custom classes with a dict() method
+    - Fallback to using __dict__ if available
+    - Recursively traverses dicts, lists, tuples, and sets
+    - Guards against circular references
+
+    Parameters
+    ----------
+    data : Any
+        The input data which may be a complex type.
+    _visited : set, optional
+        Internal set of visited object IDs to prevent infinite recursion on cyclic references.
+
+    Returns
+    -------
+    Any
+        A primitive type, or a recursively converted structure (dict, list, etc.)
+    """
+    # Initialize _visited set if not provided.
+    if _visited is None:
+        _visited = set()
+
+    # Prevent circular references by checking if the object's id has been seen.
+    data_id = id(data)
+    if data_id in _visited:
+        return data  # Return the object as-is if it has already been processed.
+    _visited.add(data_id)
+
+    # If data is a primitive type, return it directly.
+    if isinstance(data, (str, int, float, bool, type(None))):
+        return data
+
+    # Handle dataclasses by converting them to a dictionary.
+    if hasattr(data, "__dataclass_fields__"):
+        import dataclasses
+        return prepare_data(dataclasses.asdict(data), _visited=_visited)
```

Comment:
`dataclasses` is part of the Python core library, so you can import it at the top.